userd/launcher.go

```diff
@@ -93,3 +102,78 @@ func (s *Launcher) OpenURL(addr string) *dbus.Error {
 
 	return nil
 }
+
+func fdToFilename(fd int) (string, error) {
+	flags, err := sys.FcntlGetFl(fd)
+	if err != nil {
+		return "", err
+	}
+	// File descriptors opened with O_PATH do not imply access to
+	// the file in question.
+	if flags&sys.O_PATH != 0 {
+		return "", fmt.Errorf("bad file descriptor")
+	}
+
+	// Determine the file name associated with the passed file descriptor.
+	filename, err := os.Readlink(fmt.Sprintf("/proc/self/fd/%d", fd))
+	if err != nil {
+		return "", err
+	}
+
+	var fdStat, fileStat syscall.Stat_t
+	if err := syscall.Fstat(fd, &fdStat); err != nil {
+		return "", err
+	}
+	if err := syscall.Stat(filename, &fileStat); err != nil {
+		return "", err
+	}
+
+	// Sanity check to ensure we've got the right file
+	if fdStat.Dev != fileStat.Dev || fdStat.Ino != fileStat.Ino {
+		return "", fmt.Errorf("could not determine file name")
+	}
+
+	fileType := fileStat.Mode & syscall.S_IFMT
+	if fileType != syscall.S_IFREG && fileType != syscall.S_IFDIR {
+		return "", fmt.Errorf("can only open regular files and directories")
+	}
+
+	return filename, nil
+}
+
+func (s *Launcher) OpenFile(clientFd dbus.UnixFD, sender dbus.Sender) *dbus.Error {
+	// godbus transfers ownership of this file descriptor to us
+	fd := int(clientFd)
+	defer syscall.Close(fd)
+
+	filename, err := fdToFilename(fd)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	dialog, err := ui.New()
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	answeredYes := dialog.YesNo(
+		i18n.G("Allow open file?"),
+		fmt.Sprintf(i18n.G("Allow snap %q to open file %q ?"), snap, filename),
+		&ui.DialogOptions{
+			Timeout: 5 * 60 * time.Second,
+			Footer:  i18n.G("This dialog will close automatically after 5 minutes of inactivity."),
```

Comment:
That's a long time! I'm not sure what would be better though.... Perhaps add a comment on why you chose 5 minutes in case we want to revisit this again?

Reply:
The 5min is what we use in xdg-settings. It is a bit of an arbitrary choice, long enough to be not inconvenient (e.g. when a download takes long and the user does not look right away) but not so long that the dbus connection stays up forever (and auto-closes with awkward failures).